sdk/core/Azure.Core/src/RequestContent.cs

```diff
@@ -220,6 +229,107 @@ public override void Dispose()
             }
         }
 
+        private sealed class ModelSerializableContent : RequestContent
+        {
+            private SequenceWriter? _writer;
+            private BinaryData? _data;
+            private byte[]? _bytes;
+            private readonly IModelSerializable<object> _model;
+            private readonly ModelSerializerOptions _options;
+
+            public ModelSerializableContent(IModelSerializable<object> model, ModelSerializerOptions options)
+            {
+                _model = model;
+                _options = options;
+            }
+
+            public override void Dispose() => _writer?.Dispose();
+
+            private SequenceWriter GetWriter(IModelJsonSerializable<object> model)
+            {
+                if (_writer is null)
+                {
+                    var writer = new SequenceWriter();
+                    using var jsonWriter = new Utf8JsonWriter(writer);
+                    model.Serialize(jsonWriter, _options);
+                    jsonWriter.Flush();
+                    _writer = writer;
+                }
+                return _writer;
+            }
+
+            private BinaryData GetBinaryData()
+            {
+                if (_data is null)
+                {
+                    _data = _model.Serialize(_options);
+                }
+                return _data;
+            }
+
+            private byte[] GetBytes()
+            {
+                if (_bytes is null)
+                {
+                    _bytes = GetBinaryData().ToArray();
+                }
+                return _bytes;
+            }
+
+            public override void WriteTo(Stream stream, CancellationToken cancellation)
+            {
+                // a model implements both xml and json we don't know the wire format and must let the model decide.
+                if (_model is IModelJsonSerializable<object> jsonSerializable && _model is not IModelXmlSerializable<object>)
+                {
+                    GetWriter(jsonSerializable).CopyTo(stream, cancellation);
+                }
+                else if (_model is IModelXmlSerializable<object> xmlSerializable && _model is not IModelJsonSerializable<object>)
+                {
+                    using XmlWriter writer = XmlWriter.Create(stream);
+                    xmlSerializable.Serialize(writer, _options);
+                    writer.Flush();
+                }
+                else
+                {
+                    var data = GetBinaryData();
+#if NETFRAMEWORK || NETSTANDARD2_0
+                    var bytes = GetBytes();
+                    stream.Write(bytes, 0, bytes.Length);
+#else
+                    stream.Write(data.ToMemory().Span);
+#endif
+                }
+            }
+
+            public override bool TryComputeLength(out long length)
```

Comment:
Can this method throw?

Reply:
It should not throw its supposed to just return false if the length cannot be calculated.

I just realized we could use the GetBinaryData to get the size if we are in the IModelSerializable case.